### ISSUE TITLE
False negative when NamedTuple subclass redefines field in covariant manner

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#5700]
The following currently type checks without error.

```python
# pyright: strict
from typing import NamedTuple

class NT1(NamedTuple):
    number: float

class NT2(NT1):
    number: int

x1 = NT1(1.2)
x2 = NT2(1.2) # Should generate an error.
```

This is related to mypy issue https://github.com/python/mypy/issues/8663.

**Comment by erictraut** [comment]
After some exploration, I've discovered that subclassing `NamedTuple` has a bunch of unexpected (and undocumented) behaviors. If you include new fields in the subclass, they are (as you might expect) not considered part of the NamedTuple, so they don't appear in the synthesized constructor or in the iterator returned by `__iter__`.

```python
from typing import NamedTuple


class NT1(NamedTuple):
    number: float


class NT2(NT1):
    other: str = "hi"

x = NT2(1.2)
print(x.number, x.other)  # "1.2 hi"

x.other = "bye"
x.number = 1.4  # Runtime exception: can't set attribute
```

Multiple inheritance gives even stranger results:
```python
from typing import NamedTuple

class NT1(NamedTuple):
    number: float

class NT2(NamedTuple):
    other: str

class NT3(NT1, NT2):
    ...

x = NT3(1.2)

print(x.number, x.other) # "1.2 1.2" !!!!
```

Things get even weirder if the `NamedTuple` is subclassed by a dataclass as shown below:

```python
from dataclasses import dataclass
from typing import NamedTuple

class NT1(NamedTuple):
    number: float

@dataclass
class NT2(NT1):
    other: str

x = NT2("hi")

print(x.number, x.other)  # "hi hi"!!!!????
```

The implementation of `NamedTuple` is a hack that works in the simple case, but it doesn't compose well with other Python features. I'm therefore not going to attempt to model it "correctly" in pyright because it's not clear what "correct" means.
